go-controller/pkg/metrics/metrics.go

```diff
@@ -403,43 +419,80 @@ func StartMetricsServer(bindAddress string, enablePprof bool, certFile string, k
 		mux.HandleFunc("/debug/pprof/trace", pprof.Trace)
 	}
 
-	go utilwait.Until(func() {
-		var err error
-		if certFile != "" && keyFile != "" {
-			err = listenAndServeTLS(bindAddress, certFile, keyFile, mux)
-		} else {
-			err = http.ListenAndServe(bindAddress, mux)
-		}
-		if err != nil {
-			utilruntime.HandleError(fmt.Errorf("starting metrics server failed: %v", err))
+	go func() {
+		var server *http.Server
+		go utilwait.Until(func() {
+			var err error
+			if certFile != "" && keyFile != "" {
+				server = getTLSServer(bindAddress, certFile, keyFile, mux)
+				err = server.ListenAndServeTLS("", "")
+			} else {
+				server = &http.Server{
+					Addr:    bindAddress,
+					Handler: mux,
+				}
+				err = server.ListenAndServe()
+			}
+			if err != nil && err != http.ErrServerClosed {
+				utilruntime.HandleError(fmt.Errorf("starting metrics server failed: %v", err))
+			}
+		}, 5*time.Second, stopChan)
+
+		<-stopChan
+		wg.Add(1)
```

Comment:
The WaitGroup Add should be done from the same thread that does the wait, otherwise there is a race condition between them.